### DISCUSSION TITLE
My take on Argon "hit lighting"

### DISCUSSION THREAD

**Discussion by ItsShamed** [ppy/osu#21940] [Ideas] [answered]
I tried to address the hit lighting issue in the Argon Pro skin, as I feel like it's a pretty good skin to play with, and it's only a few tweaks away (including this one) from being a skin I would mainly play with.

Since the [related issue](https://github.com/ppy/osu/issues/21824) was already self-assigned, and I feel like this is a very subjective and sensitive topic, I didn't open a PR. I will use feedbacks as red or green flags to open one (or not).

This post will be updated based on the changes I will have made after feedbacks. 

> **Note**
> As mentioned in https://github.com/ppy/osu/discussions/21940#discussioncomment-4566289, most players won't notice the hit animation. However, since it can matter for spectators, we need to find balance between "enjoyability" (for the spectators) and playability (for the players).

## Comparison videos

https://user-images.githubusercontent.com/39100084/210017623-970613a7-d9c6-4d6c-b066-49f9c1b03ccf.mp4

https://user-images.githubusercontent.com/39100084/210018781-a89e859a-feac-4009-a90d-bb4b4b834f1a.mp4

The goal was to make the hit animation more subtle and less distracting, while keeping the feeling of the original animation. Essentially, this will improve the reading experience with this skin, as it will no longer block stacked objects in sight.

### Changes (for those who can't or don't want to watch the videos)

- Obviously, the `FlashPiece` has been discarded completely.
- The global animation duration went from 800 to 240. This number comes from the `LegacyMainCirclePiece` fade duration. Since most players coming from `LegacySkin`s are used to this duration, it seemed obvious to me to use this value: it's long enough to notice it (doesn't disappear brutally) and it's short enough to not distract.
- The `outerGradient` no longer turns white, but instead fades out at the same time it's being shrunk. This is to keep consistency with the hit object and its color, otherwise it looks like you're playing inside the editor.
- Most animations now use `Easing.Out` (except for the global fade out, which is linear): at high speed `Easing.OutQuint` and `Easing.OutQuad` look like the animation is instant and `Easing.OutElasticHalf` is too distracting in my opinion.
> **Note**
> The listed changes are all controlled by a single boolean and the original animation (as it was) is kept when the "Hit Lighting" setting is enabled or the skin is the normal Argon skin (although this last part of the condition may change later, depending on feedback).

### [`d9c8b31`](https://github.com/ppy/osu/commit/d9c8b31cf58796abe1381b1633ffb2dc21b10ee8) -- Attempt at bringing more impact

https://user-images.githubusercontent.com/39100084/210174339-219c9862-9aaf-49a5-9c28-36eb4ecfddf2.mp4

- Made `resize_duration` const again (with its original value)
- For resized components, they now use `OutElasticHalf` as before

**Comments:** Some Quality of Code issues were left so that if refactoring were to be done, I would not have to revert too many commits and continue to work on this.

### [`3efbdca`](https://github.com/ppy/osu/commit/3efbdca81b548a6aa7eeb737c68f198481be365f) -- Second pass at bringing more impact

https://user-images.githubusercontent.com/39100084/210413326-99fc0c89-62a8-412d-8f4c-dfc13b2d6824.mp4

https://user-images.githubusercontent.com/39100084/210443141-8c82db8e-3413-486e-831f-108cf2aefddc.mp4

- `outerGradient` flashes to white briefly. Flashing colors can give a feeling of impact.

**Comments:** This will need more in-game testing and more player feedback.

---

You can test this by yourself on [this branch (`skin/argon-hit-lighting`)](https://github.com/itsshamed/osu/tree/skin/argon-hit-lighting).

**Comment by peppy** [comment]
The hit lighting off version lacks any kind of impact on hitting. It doesn't look like it would feel satisfying to play with.

**Reply by ItsShamed** [reply]
Hmm, maybe I should revert some easing changes then, that was kind of the goal to make the animation "smooth"

**Reply by peppy** [reply]
There's also another alternative at https://github.com/ppy/osu/pull/21981 which was abruptly PR'd, which should probably be considered.

**Reply by ItsShamed** [reply]
Oh yeah, I'll try to play with that to see if I'm comfortable with it, if not I'll try to continue working on my side.

Update (warning: really subjective opinion), just play tested the branch: while it is comfortable to play with because it's basically a port of the whole legacy animation onto the Argon skin, it might not fit the design of Argon (because it lacks of personality), which was the goal of my implementation (keep the feeling of the original Argon animation).

**Reply by yamplum** [reply]
I investigated some of the skins used by actual top players, and it seems that they mostly stick to the default "grow + fade out" animation for visual feedback. The important thing to me seems that the afterglow is never brighter than the actual circle, which both triangles and argon suffer from.

EDIT: also, subjectively, I don't find hit impact to be very important. People play Hidden just fine and you don't see the circles there at all. I'd love to see argon pro focus on minimizing distractions.

Whitecat:

https://user-images.githubusercontent.com/118788705/210177559-441ee8b3-b742-4423-a967-40500352453d.mp4

BTMC:

https://user-images.githubusercontent.com/118788705/210177565-63b4f4e3-3ee0-47c3-8639-ec53b9aee8d5.mp4

RyuK:

https://user-images.githubusercontent.com/118788705/210177572-02f22a08-feaf-4198-b544-b20500c41ec9.mp4

My personal use skin:

https://user-images.githubusercontent.com/118788705/210177581-994cf984-0017-4e36-82eb-144f27084435.mp4

**Reply by frenzibyte** [reply]
It's not like you can change the hit animation in stable, can you? All skins on stable will have the same animation.

**Reply by yamplum** [reply]
I suppose, but it's worth nothing that hit lighting is also disabled, so I think the design should proceed with that in mind. Stable skins have had a long time to develop and have served players well, so it's worth examining them when making diverging changes.

**Reply by ItsShamed** [reply]
> they mostly stick to the default "grow + fade out"

As @frenzibyte said, this is a default behavior of stable (Legacy Skins on lazer), which is why I find that reimplementing the legacy animation on Argon feels really weird to think about.

> I don't find hit impact to be very important

I don't either, but it's true that impact gives more character. My statement about `OutElasticHalf` seems kinda wrong now, since I managed to make it look good.

> Stable skins have had a long time to develop and have served players well, so it's worth examining them when making diverging changes

If you read the changes made in the original post, I did exactly that, leading me to choose 240 as the fade duration.

**Reply by yamplum** [reply]
I've tried out the updated version. It's not distracting, so I'm happy with it. As far as different easing types go, it seems to be mostly a concern for spectators -- I found that as a player I always look 1-2 circles ahead, so I don't notice the afterhits much. The only time they're obvious is on longer streams, but they cause no issues and I find this version very playable overall. Thanks for taking the time.

**Reply by peppy** [reply]
@ItsShamed could you please PR the current version of your changes? I think it's a starting point and I'll fine-tune from there if required.

**Comment by yuuiko** [comment]
Yeah with them just disappearing theres no impact to feel

I like the bright flash in argon current
Perhaps for pro just make the fadeout faster but keep the same bright flash

**Reply by ItsShamed** [reply]
While I think many won't agree with you because the animation being too bright is distracting to the player eye during gameplay, I do agree that it needs more impact. I'm currently looking into this, and I'll update the post if changes are being made. 

(The post has been updated, please take a look. Also, don't hesitate to build the branch and test by yourself)

**Reply by yuuiko** [reply]
I think the problem isn't the bright flash but instead how long the afterimage stays around

**Reply by ItsShamed** [reply]
> I think the problem isn't the bright flash but instead how long the afterimage stays around

The fade duration is already short enough, though... And I just tried bringing back the bright flash, it doesn't feel right imo.

**Comment by camellirite** [comment]
i don't have visual studio installed so i can't check in-game but the video looks good

**Comment by ItsShamed** [comment, marked as answer]
Tracking via #22392
